Added code to assign turkers to a specific sequence of routes

maddalihanumateja opened the pull request:
I've added the following features:
- Turkers will now be assigned to a condition present in the amt_condition table
- Each condition is associated with a set id (hardcoded in this branch with the name volunteer_id). This restricts the turkers to work on a restricted set of routes. In this case, the set of routes associated with missions performed by volunteers on the sidewalk website.
- A maximum of 5 turkers can be assigned a single condition. The condition assigned to a turker is the one with the least number of allottees at that time. 
- This branch assumes that a HIT is presented as completing a single route. When turkers completes a route  they submit the HIT and exit. They need to accept another HIT to work on the next route associated with the condition. I am currently working on a different branch that will present multiple routes in the same HIT session.

Tables added/modified:
- The turker table has been modified to include condition assigned to the turker.
- The amt_condition table has been modified to include a volunteer_id field (a unique identifier for a set of routes. This should be renamed to something like route_set_id in the future)
- The amt_volunteer_route table stores mapping between volunteer_id (of the turker and amt_condition tables) and route_id (of the route and route_street tables)

For testing:
- The python file used for generating the amt_condition, amt_volunteer_route needs some cleanup. What you'll be working with is a sql dump for the database.
- The workflow to audit routes is similar to how it was being done using the previous mturk repo. 
    - Delete the sidewalk database on your machine and restore from the sql dump ([link](https://www.dropbox.com/s/ztk5hzyuafqsuky/sidewalk-mturk-test-dump?dl=0)) using pg_restore
    - Visit the url http://localhost:9000/?assignmentId=some_string&workerId=some_string&hitId=some_string&turkSubmitTo=localhost:5000/post . This will automatically assign you to a condition id.
    - Check that the turker table inserts a new record with your workerId and conditionId
    - Check that a record is inserted in the amt_assignment table with a route_id that is correctly associated with the assigned condition_id.
    - Check that the distance to be audited in the mission is 1000ft for the first and second missions and 2000ft (displayed as 1/2 mile for some reason) for the third mission.
    - Finish the mission associated with the route and click on the Submit HIT button. 
    - Reload the audit page with a different assignmentId and hitId. In the amt_assignment table you should check that the route_id of this second mission is the second route associated with the condition assigned to workerId.
    - Continue auditing routes, submitting them, and revisiting the page with updated assignmentId and hitId till you are routed to the main page. This indicates that you have completed all routes that were available to you. (This is a temporary measure to handle users who have worked on all routes associated with their assigned condition)

misaugstad commented:
After going to the URL, entries have been added to all the tables, like you said, but then I just sit on a loading screen, and don't get to audit. Console error shown below.

![loading-routemodelreferror](https://user-images.githubusercontent.com/6518824/29255009-5c57ab6c-806b-11e7-8424-d8992a88d960.png)
![loading-code](https://user-images.githubusercontent.com/6518824/29255011-5f30acee-806b-11e7-8af1-8737123c5b03.png)


I did make sure to do this incognito this time to make sure that I am not logged in!

maddalihanumateja commented:
This is pre-existing code. Never saw this error before so I'm unsure of how to replicate it. I'm assuming you did a grunt rerun and tried reloading the page. Does your sv_label.js file even have a RouteModel object in it? Also, please recheck the repo you are using.

misaugstad commented:
1. I don't know what a grunt rerun is. If you just mean that I reran `grunt watch` then yeah I did that.
2. Yep my SVLabel.js file has RouteModel all up in it.
3. Yep I am using the correct repo.

I just cloned the entire repo again. My grunt watcher thing isn't working as expected, so it didn't update the build files. I have a mission now!

misaugstad commented:
On my 2nd try, I was able to complete all 3 HITs!

1. If a mission finished in the middle of a street, the next mission seemed to start me over at the beginning of that street. Maybe that was just a specific scenario where that happened, since it could happen if the user originally left the audit page after their first mission and then came back.
2. During my 2nd mission, my completion percentage said 75%, but my next step brought up the mission complete screen. Then in my 3rd mission, the completion said 55%, then the next step said mission complete.
3. In route 12, I got to a point where the route stopped after just the 2nd street segment. It didn't say that I was at the end of the route and needed to jump, either. Even tried refreshing, the route just was broken :( No console errors.
![nomoresegs](https://user-images.githubusercontent.com/6518824/29256248-c4bacf9e-8076-11e7-9694-6620c1a407de.png)
4. Same route, interesting that it wants me to U-turn...
![whyuturn](https://user-images.githubusercontent.com/6518824/29256273-1c43de04-8077-11e7-9ebd-57d402f9824f.png)
5. It seems that when the 2nd HIT is first started, it says you have not audited any distance in this neighorborhood... But once you do a jump, the distance you had audited in the first HIT shows up in the top-right where it says how much you have audited in this neighborhood. Pictured below, pre and post jumping.
![completion-prejump](https://user-images.githubusercontent.com/6518824/29256370-eb2cb952-8077-11e7-8431-63ced4ca8bd4.png)
![completion-postjump](https://user-images.githubusercontent.com/6518824/29256371-ed06df0a-8077-11e7-9c72-fcde53ccdb84.png)
6. When I try to do a fourth HIT, you said that it should bring me to the main page, but instead I get this:
![fourthhit](https://user-images.githubusercontent.com/6518824/29256445-606acefc-8078-11e7-9615-57ba12ccda67.png)

maddalihanumateja commented:
I think that you are working an older version of the branch. Did you pull from the repo just to see if it was up to date?

misaugstad commented:
I was working on an up-to-date version of the branch. I fixed the error and pushed the fix. Now more comments as I test out some more.

7. I got this error when loading the page once. Upon inspection, it is because this route has only 1 street. In the route table in the database, you allow the std_street_length_mi column to be null, but in Route class that is defined, this column isn't an Option[Double], it is a Double. I fixed this and am pushing the commit as well.
![resultseterror](https://user-images.githubusercontent.com/6518824/29278168-eee1495e-80e1-11e7-9bf1-fc54a943b207.png)

8. It seems that the distance that the user sees they have audited is not updated until they jump. See below. At the end of the first mission, distance still says 0.0, but 1000ft is 0.19 miles. Then when I complete the mission, it says 0.2 miles. NOTE, later on I was going through and the distance was being updated after finishing an audit task, not sure what is changing.
![distnotupdated](https://user-images.githubusercontent.com/6518824/29278982-70075bac-80e4-11e7-9193-756cd0a9bb71.png)

9. At some point in the middle of my 2nd mission I had to jump. After my jump, it said I had completed 0.5 miles in the neighborhood, but 2000 is only 0.38 miles so it definitely shouldn't say that. Then after a bit it went back down to 0.3 miles. I'm not sure what made it go back down, when it happened I simultaneously double clicked to move forward, and finished an audit task. And I am in incognito, so it isn't distance carried over from previous routes I had done.

10. Yeah that distance being shown is all messed up. Third mission, same neighborhood, I start with 0.0 miles, finish an audit task, it says 0.2 miles now (and the task was wayyy less than that, and also I should be at about 0.5 total now. Jumped, and it went to 0.4. When I finished the last mission, the total came up to 0.5, but 4000ft is 0.76 miles.

11. Although this time the progress bar was correct for the first 2 missions, on the 3rd mission the mission ended when the progress bar was at about 50%. Then I went through again, and 2nd mission ended at 60%. So it is all happening incorrectly only sometimes...

12. Found another case of being told to do a U-turn, and no new street showing up. I looked at this particular route in your dump (route 172), and it seems that it consists of just the same street edge, twice in a row!
![uturnagain](https://user-images.githubusercontent.com/6518824/29279623-6c6a478c-80e6-11e7-9661-34a959993677.png)
![doublestreet](https://user-images.githubusercontent.com/6518824/29279674-a1cde0dc-80e6-11e7-89a8-8454b038f488.png)

13. I then checked route 12, the one that gave me the same issue as I commented previously, and the issue seems to be that an entry in the route_street table has the current_street_edge_id equal to the next_street_edge_id.

Final thoughts...
Issue that *needs* to be addressed: Routes with current_street_edge_id equal to the next_street_edge_id for some edge in the route.

Issue that *should* be addressed: progress bar showing incorrect completion percentage.

Issue that doesn't need to be addressed right now, but should definitely be addressed before sending it out to turkers: Distance completed in neighborhood being incorrect.

misaugstad commented on app/models/route/RouteTable.scala:
```diff
@@ -43,6 +43,10 @@ object RouteTable{
     route.headOption
   }
+  def getRegionByRouteId(routeId: Option[Int]): Option[Int] = db.withSession { implicit session =>
+    val regionId = routes.filter(_.routeId === routeId).list.map(_.regionId).headOption
```
You can select a single column using slick instead of doing select * then filtering.
routes.filter(_.routeId === routeId).map(_.regionId).list.headOption

maddalihanumateja replied:
changed it

misaugstad commented on app/models/turker/TurkerTable.scala:
```diff
@@ -26,6 +31,18 @@ object TurkerTable{
+  }
+
+  def getConditionIdByTurkerId(turkerId: String): Option[Int] = db.withTransaction { implicit session =>
+    val cId = turkers.filter(_.turkerId === turkerId).list.headOption
```
Try to use Scala's strong typing in all the code you write. Also rename this variable....
val turker: Turker = turkers.filter...

maddalihanumateja replied:
changed it

misaugstad commented on app/models/turker/TurkerTable.scala:
```diff
@@ -26,6 +31,18 @@ object TurkerTable{
+  def getConditionIdByTurkerId(turkerId: String): Option[Int] = db.withTransaction { implicit session =>
+    val cId = turkers.filter(_.turkerId === turkerId).list.headOption
+    cId match {
+      case Some(condition) => Some(condition.amtConditionId)
```
And rename this variable.

maddalihanumateja replied:
changed it

misaugstad commented on app/models/mission/MissionTable.scala:
```diff
@@ -216,6 +216,11 @@ object MissionTable {
   }
 
+  def selectMTurkMissionByRegion(regionId: Option[Int]): List[Mission] = db.withSession { implicit session =>
+    missionsWithoutDeleted.filter(_.label inSet List("mturk-mission")).filter(_.regionId === regionId).list ++ missionsWithoutDeleted.filter(_.label inSet List("onboarding")).list
```
missionsWithoutDeleted.filter(l => (l.label === "mturk-mission" && l.regionId === regionId) || l.label === "onboarding).list

maddalihanumateja replied:
Thanks. Dont know why this didnt strike me.

misaugstad commented on app/models/mission/MissionTable.scala:
```diff
@@ -216,6 +216,11 @@ object MissionTable {
 
+  def selectMTurkMissionByRegion(regionId: Option[Int]): List[Mission] = db.withSession { implicit session =>
+    missionsWithoutDeleted.filter(_.label inSet List("mturk-mission")).filter(_.regionId === regionId).list ++ missionsWithoutDeleted.filter(_.label inSet List("onboarding")).list
+    //missionsWithoutDeleted.filter(l => l.label === "mturk-mission" || l.label ===  "onboarding").list
```
remove this comment

misaugstad commented on app/models/amt/AMTAssignmentTable.scala:
```diff
@@ -66,6 +66,10 @@ object AMTAssignmentTable {
     q.update(completed)
   }
 
```
This function should be placed before all the saving/updating functions (to match conventions of the rest of the codebase). It could also probably use a brief comment to say what it does more explicitly.

maddalihanumateja replied:
This is old code. I've added the comment.

misaugstad commented on app/models/amt/AMTConditionTable.scala:
```diff
@@ -28,6 +33,29 @@ class AMTConditionTable(tag: Tag) extends Table[AMTCondition](tag, Some("sidewal
   val db = play.api.db.slick.DB
   val amtConditions = TableQuery[AMTConditionTable]
+  val amtAssignments = TableQuery[AMTAssignmentTable]
+  val max_num_condition_assignments = 5
```
Put this in camelCase and specify data type
val maxNumConditionAssignments: Int = 5

maddalihanumateja replied:
changed it

misaugstad commented on app/models/amt/AMTConditionTable.scala:
```diff
@@ -28,6 +33,29 @@ class AMTConditionTable(tag: Tag) extends Table[AMTCondition](tag, Some("sidewal
+  val amtAssignments = TableQuery[AMTAssignmentTable]
+  val max_num_condition_assignments = 5
+
+  def getVolunteerIdByConditionId(amtConditionId: Int): String = db.withTransaction { implicit session =>
```
This should probably be an Option[String] since the amtConditionId might not be in the database.

misaugstad commented on app/models/amt/AMTConditionTable.scala:
```diff
@@ -28,6 +33,29 @@ class AMTConditionTable(tag: Tag) extends Table[AMTCondition](tag, Some("sidewal
+  val max_num_condition_assignments = 5
+
+  def getVolunteerIdByConditionId(amtConditionId: Int): String = db.withTransaction { implicit session =>
+    val vId = amtConditions.filter(_.amtConditionId === amtConditionId).list.headOption.map(_.volunteerId)
```
Can do "select volunteer_id" instead of "select *"...
amtConditions.filter(\_.amtConditionId === amtConditionId).map(\_.volunteerId).list.headOption

maddalihanumateja replied:
changed it

misaugstad commented on app/models/amt/AMTConditionTable.scala:
```diff
@@ -28,6 +33,29 @@ class AMTConditionTable(tag: Tag) extends Table[AMTCondition](tag, Some("sidewal
+  def assignAvailableCondition: Option[Int] =  db.withTransaction { implicit session =>
+    //Get the condition id with the least number of current assignments
+
+    val selectConditionId = Q.query[Int, Int](
```
rename this selectConditionIdQuery

misaugstad commented on app/models/amt/AMTConditionTable.scala:
```diff
@@ -28,6 +33,29 @@ class AMTConditionTable(tag: Tag) extends Table[AMTCondition](tag, Some("sidewal
+      """.stripMargin
+    )
+
+    val cId = selectConditionId(max_num_condition_assignments).list.headOption
```
do val cId: Int = ... or just don't save it in a val before returning.

misaugstad commented on app/controllers/MissionController.scala:
```diff
@@ -107,6 +110,31 @@ class MissionController @Inject() (implicit val env: Environment[User, SessionAu
+            "distance_ft" -> m._1.distance_ft,
+            "distance_mi" -> m._1.distance_mi,
+            "coverage" -> m._1.coverage)
+    )
```
fix indentation

maddalihanumateja replied:
changed it

misaugstad commented on app/controllers/MissionController.scala:
```diff
@@ -107,6 +110,31 @@ class MissionController @Inject() (implicit val env: Environment[User, SessionAu
+  def getMTurkMissionsByTurker(turkerId: String) = UserAwareAction.async { implicit request =>
+    request.identity match {
+      case _ =>
+        val conditionId = TurkerTable.getConditionIdByTurkerId(turkerId).get
```
declare data types of all these variables (e.g., val conditionId: Int = ...)

misaugstad commented on app/controllers/AuditController.scala:
```diff
@@ -109,31 +112,48 @@ class AuditController @Inject() (implicit val env: Environment[User, SessionAuth
-            // Retrieve the route based on HIT ID
-            val routeId: Option[Int] = AMTRouteAssignmentTable.findRouteByHITId(hitId)
+            // Retrieve the route based on the condition that the worker has been assigned to and the associated unvisited routes
+            var cId = AMTConditionTable.assignAvailableCondition
```
Instead of making cId a var and then doing if/else, switch this to Scala's match syntax...
```
val cId: Int = TurkerTable.getConditionIdByTurkerId(workerId) match {
    case Some(conditionId) =>
        TurkerTable.getConditionIdByTurkerId(workerId)
    case None =>
        val newCId: Int = AMTConditionTable.assignAvailableCondition
        val turker: Turker = Turker(workerId, "",newCId.getOrElse(1))
        TurkerTable.save(turker)
        newCId
}
```

maddalihanumateja replied:
done.

misaugstad commented on app/controllers/AuditController.scala:
```diff
@@ -109,31 +112,48 @@ class AuditController @Inject() (implicit val env: Environment[User, SessionAuth
+                val asgId: Option[Int] = Option(AMTAssignmentTable.save(asg))
+
+                // Load the first task from the selected route
+                val regionId = theRoute.regionId
```
val regionId: Int = ...

misaugstad commented on app/controllers/AuditController.scala:
```diff
@@ -109,31 +112,48 @@ class AuditController @Inject() (implicit val env: Environment[User, SessionAuth
+                // Load the first task from the selected route
+                val regionId = theRoute.regionId
+                val region: Option[NamedRegion] = RegionTable.selectANamedRegion(regionId)
+                val regionName = region.get.name
```
val regionName: String = ...

misaugstad commented on app/models/amt/AMTVolunteerRouteTable.scala:
```diff
@@ -0,0 +1,73 @@
+  *
+  */
+class AMTVolunteerRouteTable(tag: Tag) extends Table[AMTVolunteerRoute](tag, Some("sidewalk"), "amt_volunteer_route") {
+  def amtVolunteerRouteId = column[Int]("amt_volunteer_route_id", O.NotNull, O.PrimaryKey)
```
Shouldn't this be auto incrementing?

maddalihanumateja replied:
No

misaugstad commented on app/models/amt/AMTVolunteerRouteTable.scala:
```diff
@@ -0,0 +1,73 @@
+  val amtVolunteerRoutes = TableQuery[AMTVolunteerRouteTable]
+  val amtAssignments = TableQuery[AMTAssignmentTable]
+
+  def findRoutesByVolunteerId(volunteerId: String): List[Int] = db.withTransaction { implicit session =>
```
rename to findRouteIdsByVolunteerId

maddalihanumateja replied:
changed it

misaugstad commented on app/models/amt/AMTVolunteerRouteTable.scala:
```diff
@@ -0,0 +1,73 @@
+  val amtAssignments = TableQuery[AMTAssignmentTable]
+
+  def findRoutesByVolunteerId(volunteerId: String): List[Int] = db.withTransaction { implicit session =>
+    val routeAsg = amtVolunteerRoutes.filter(_.volunteerId === volunteerId).map(_.routeId).list
```
don't need to save in variable

misaugstad commented on app/models/amt/AMTVolunteerRouteTable.scala:
```diff
@@ -0,0 +1,73 @@
+    val routeAsg = amtVolunteerRoutes.filter(_.volunteerId === volunteerId).map(_.routeId).list
+    routeAsg
+  }
+
```
This function is complex enough to deserve a comment before the method.

maddalihanumateja replied:
I don't think so. Do you have any specific questions about this function? It seems simple enough.

misaugstad commented on app/models/amt/AMTVolunteerRouteTable.scala:
```diff
@@ -0,0 +1,73 @@
+  def assignRouteByConditionIdAndWorkerId(conditionId: Int, workerId: String): Option[Int] = db.withTransaction { implicit session =>
+    // Find the first route in the list of routes associated with volunteerId (these are obtained using findRoutesByVolunteerId)
+    // that hasnt been audited by workerId (this can be checked in the amt_assignment table)
+    val volunteerId = AMTConditionTable.getVolunteerIdByConditionId(conditionId)
```
val volunteerId: Int = ... 
etc

misaugstad commented on app/models/amt/AMTVolunteerRouteTable.scala:
```diff
@@ -0,0 +1,73 @@
+    // that hasnt been audited by workerId (this can be checked in the amt_assignment table)
+    val volunteerId = AMTConditionTable.getVolunteerIdByConditionId(conditionId)
+    val availableRoutes = findRoutesByVolunteerId(volunteerId)
+    val auditedRoutes = amtAssignments.filter(_.turkerId === workerId).filter(_.completed).map(_.routeId).list.map(route => route.get)
```
Instead of list.map(route => route.get), you can do list.flatten
So I would write this line as
```
val auditedRoutes: List[Int] = amtAssignments.filter(asmt =>
    asmt.completed && asmt.turkerId === workerId).map(_.routeId).list.flatten
```

maddalihanumateja commented on app/controllers/AuditController.scala:
```diff
@@ -109,31 +112,48 @@ class AuditController @Inject() (implicit val env: Environment[User, SessionAuth
+
+            if(TurkerTable.getConditionIdByTurkerId(workerId) == None){
+              // No worker was found with this id (implies no condition was assigned) so assign the condition id that has been least used
+              // Select amt_condition.amt_condition_id, count(condition_id) as cnt from amt_assignment Right JOIN amt_condition on (amt_assignment.condition_id = amt_condition.amt_condition_id) group by amt_condition.amt_condition_id order by cnt asc;
```
That was the query I used to test if I was getting the correct condition id. I've removed the comment.

maddalihanumateja commented on app/controllers/AuditController.scala:
```diff
@@ -109,31 +112,48 @@ class AuditController @Inject() (implicit val env: Environment[User, SessionAuth
+              val turker: Turker = Turker(workerId, "",cId.getOrElse(1))
+              TurkerTable.save(turker)
+
+              // Bug: turker id is taken as null
```
It doesnt now. The bug is resolved.

maddalihanumateja commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionComplete.js:
```diff
@@ -128,6 +128,7 @@ function ModalMissionComplete (svl, missionContainer, taskContainer,
 
     //uiModalMissionComplete.background.on("click", this._handleBackgroundClick);
     //uiModalMissionComplete.closeButton.on("click", this._handleCloseButtonClick);
+    //Add code here to make missions appear consecutively on a single HIT rather on multiple HITs
```
Done

maddalihanumateja commented on public/javascripts/SVLabel/src/SVLabel/mission/MissionProgress.js:
```diff
@@ -130,9 +141,13 @@ function MissionProgress (svl, gameEffectModel, missionModel, modalModel, neighb
         var currentNeighborhoodId = currentNeighborhood.getProperty("regionId");
         var nextMission = missionContainer.nextMission(currentNeighborhoodId);
 
+        //Add code here to bring up the submit HIT button and post to the turkSubmit link
```
Done

maddalihanumateja commented on public/javascripts/SVLabel/src/SVLabel/mission/MissionProgress.js:
```diff
@@ -141,8 +156,9 @@ function MissionProgress (svl, gameEffectModel, missionModel, modalModel, neighb
         }
 
         // Adjust the target distance based on the tasks available
+        //Need to check if this is require for mturk code
```
Done

maddalihanumateja commented on hit_creation_sim.py:
```diff
@@ -37,6 +37,26 @@ def create_missions_for_routes(engine, cursor, route_rows):
+                'label': 'mturk-mission',
+                'level': 2,
+                'deleted': False,
+                'coverage': None,
```
I need to confirm whether the progress bar even interacts with the coverage attribute for this

maddalihanumateja commented on conf/evolutions/default/5.sql:
```diff
@@ -0,0 +1,27 @@
+  ADD volunteer_id TEXT NOT NULL;
+
+CREATE TABLE amt_volunteer_route(
+  amt_volunteer_route_id INT,
```
Nope. This isn't an auto increment field. I'm generating this in python.

maddalihanumateja commented on conf/evolutions/default/5.sql:
```diff
@@ -0,0 +1,27 @@
+
+ALTER TABLE turker
+  DROP amt_condition_id;
+  DROP CONSTRAINT IF EXISTS turker_amt_condition_id_fkey;
```
Never got an error. Probably because of the "if exists" condition

maddalihanumateja commented:
I've addressed the code review suggestions as much as possible in the latest commits. 

 - Routes with current_street_edge_id equal to the next_street_edge_id for some edge in the route (I need to look into route generation for this to see what is causing the problem)
 -  Progress bar showing incorrect completion percentage. (Unknown reason. I'll try to figure this out by tomorrow)
 -  Distance completed in neighborhood is incorrect. (There were no changes to the javascript that handles either the mission progress bar or distance calculation on this branch. It might be an artifact from old code. Currently unknown reason)
